samza-core/src/main/java/org/apache/samza/operators/impl/store/TimeSeriesStoreImpl.java

```diff
@@ -0,0 +1,200 @@
+ *     with an existing key and timestamp will append the value to the list.
+ *   </li>
+ * </ol>
+ * <p>
+ * Implementation Notes:
+ *
+ * Data is serialized and organized into K-V pairs as follows:
+ *  <pre>
+ *    +-----------------------+------------------+------------+------------------------+-----------------------+
+ *    |  serialized-key bytes |  timestamp       | version    |       sequence number  | serialized value      |
+ *    |                       |                  |            |                        |                       |
+ *    +-----------------------+------------------+------------+------------------------+-----------------------+
+ *    +----------------------+--------8 bytes----+----1 bytes-+---------7 bytes--------+----value size----------
+ *    +----------------------------------STORE KEY-------------------------------------+---STORE VAL-----------+
+ *  </pre>
+ *  An 8 byte timestamp, a one byte version and a 7 byte sequence number are appended to the provided key and this
+ *  combination is used as the key in the k-v store. The provided value is stored as is.
+ *
+ *  <p> This class is thread-safe and concurrent reads/writes are expected.
+ *
+ * @param <K> the type of key in the store
+ * @param <V> the type of value in the store
+ */
+public class TimeSeriesStoreImpl<K, V> implements TimeSeriesStore<K, V> {
+
+  private final KeyValueStore<TimeSeriesKey<K>, V> kvStore;
+
+  /**
+   * Since timestamps are at the granularity of milliseconds, multiple entries added in the same
+   * millisecond are distinguished by a monotonically increasing sequence number.
+   */
+  private final AtomicLong seqNum = new AtomicLong();
+  private final boolean appendMode;
+
+  /**
+   * Creates a {@link TimeSeriesStoreImpl}
+   *
+   * @param kvStore the backing kv store to use
+   * @param appendMode should the store be used in appendMode
+   */
+  public TimeSeriesStoreImpl(KeyValueStore<TimeSeriesKey<K>, V> kvStore, boolean appendMode) {
+    this.kvStore = kvStore;
+    this.appendMode = appendMode;
+  }
+
+  /**
+   * Creates a {@link TimeSeriesStoreImpl} in append mode.
+   *
+   * @param kvStore the backing kv store to use
+   */
+  public TimeSeriesStoreImpl(KeyValueStore<TimeSeriesKey<K>, V> kvStore) {
+    this(kvStore, true);
+  }
+
+  @Override
+  public void put(K key, V val, long timestamp) {
+    // For append mode, values are differentiated by an unique sequence number. For overwrite mode, the sequence
+    // number is always zero. This ensures that only the most recent value is retained.
+    if (appendMode) {
+      seqNum.getAndIncrement();
+    }
+    TimeSeriesKey<K> timeSeriesKey = new TimeSeriesKey<>(key, timestamp, seqNum.get());
+    kvStore.put(timeSeriesKey, val);
+  }
+
+  @Override
+  public ClosableIterator<TimestampedValue<V>> get(K key, long startTimestamp, long endTimestamp) {
```

Comment:
Why not close the iterator automatically when it reaches to the end? Now we have to remind user to close it after using.

Reply:
- Closing iterators automatically means that `iterator.hasNext()` is stateful - The invocation of `hasNext` immediately after the iterator has been closed will return an `InvalidStateException` . 
- Alternately, If we choose to return `false` during `hasNext()`, we should have a way to distinguish these 2 cases:
* `false` returned due to no elements to iterate upon
* `false` returned due to iterator being closed.

For these reasons, and to maintain consistency with our already existing iterator APIs that are built on top of RocksDb, I'd prefer to avoid auto-closing iterators.